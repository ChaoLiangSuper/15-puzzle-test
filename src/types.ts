export interface IAppState {
  tiles: Number[];
  isSolved: boolean;
}

export interface ICellProp {
  displayValue: number;
  handleCellclick: () => void;
}

export interface IGridProp {
  tiles: number[];
  edgeLength: number;
  handleCellclick: (p: number) => void;
  isSolved: boolean;
}

export interface IOverlayProps {
  displayText: string;
}
