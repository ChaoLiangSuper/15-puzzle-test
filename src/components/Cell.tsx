import React from 'react';
import { ICellProp } from '../types';

const Cell: React.FC<ICellProp> = ({ displayValue, handleCellclick }) => {
  const className = displayValue === 0 ? 'cell cell-empty' : 'cell';
  return (
    <div className={className} onClick={handleCellclick}>
      {displayValue === 0 ? null : displayValue}
    </div>
  );
};

export default Cell;
