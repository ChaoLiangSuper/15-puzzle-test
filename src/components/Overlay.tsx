import React from 'react';
import { IOverlayProps } from '../types';

const Overlay: React.FC<IOverlayProps> = ({ displayText }) => (
  <div className='overlay'>{displayText}</div>
);

export default Overlay;
