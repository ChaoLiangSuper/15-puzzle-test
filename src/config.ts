export const edgeLength: number = 4;
