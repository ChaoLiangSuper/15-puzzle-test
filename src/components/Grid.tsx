import _ from 'lodash';
import React from 'react';
import Cell from './Cell';
import Overlay from './Overlay';
import { IGridProp } from '../types';

const Grid: React.FC<IGridProp> = ({
  tiles,
  edgeLength,
  handleCellclick,
  isSolved,
}) => {
  const cells = _.map(tiles, (value, index) => (
    <React.Fragment key={index}>
      <Cell
        displayValue={value}
        handleCellclick={() => handleCellclick(index)}
      />
      {(index + 1) % edgeLength === 0 ? <div /> : null}
    </React.Fragment>
  ));

  return (
    <div className='grid'>
      {cells}
      {isSolved ? <Overlay displayText='WIN' /> : null}
    </div>
  );
};

export default Grid;
