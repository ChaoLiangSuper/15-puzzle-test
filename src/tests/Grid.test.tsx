import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Grid from '../components/Grid';
import Cell from '../components/Cell';
import Overlay from '../components/Overlay';
import { IGridProp } from '../types';
import { edgeLength } from '../config';
import { mockTiles } from './mock';

describe('<Grid />', () => {
  let wrapper: ShallowWrapper<IGridProp>;
  let props: IGridProp;

  beforeEach(() => {
    props = {
      tiles: mockTiles,
      edgeLength,
      handleCellclick: jest.fn(),
      isSolved: false,
    };
    wrapper = shallow(<Grid {...props} />);
  });

  it('should initialize without crash and match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render <Cell /> based on received tiles', () => {
    const cells = wrapper.find(Cell);
    expect(cells.length).toBe(mockTiles.length);
    expect(cells.map(cell => cell.props().displayValue)).toEqual(mockTiles);
  });

  it('should render <Overlay /> if the game if solved', () => {
    wrapper.setProps({
      ...props,
      isSolved: true,
    });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Overlay).length).toBe(1);
  });

  it('should handle cell click', () => {
    expect(props.handleCellclick).toHaveBeenCalledTimes(0);

    const cells = wrapper.find(Cell);
    cells
      .at(0)
      .props()
      .handleCellclick();
    expect(props.handleCellclick).toHaveBeenCalledTimes(1);
    expect(props.handleCellclick).toHaveBeenCalledWith(0);

    cells
      .at(5)
      .props()
      .handleCellclick();
    expect(props.handleCellclick).toHaveBeenCalledTimes(2);
    expect(props.handleCellclick).toHaveBeenCalledWith(5);
  });
});
