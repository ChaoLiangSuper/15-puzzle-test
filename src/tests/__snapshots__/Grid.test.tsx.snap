// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> should initialize without crash and match snapshot 1`] = `
<div
  className="grid"
>
  <Cell
    displayValue={1}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={2}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={3}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={4}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={5}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={6}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={7}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={8}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={9}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={10}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={11}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={12}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={13}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={14}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={0}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={15}
    handleCellclick={[Function]}
  />
  <div />
</div>
`;

exports[`<Grid /> should render <Overlay /> if the game if solved 1`] = `
<div
  className="grid"
>
  <Cell
    displayValue={1}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={2}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={3}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={4}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={5}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={6}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={7}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={8}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={9}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={10}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={11}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={12}
    handleCellclick={[Function]}
  />
  <div />
  <Cell
    displayValue={13}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={14}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={0}
    handleCellclick={[Function]}
  />
  <Cell
    displayValue={15}
    handleCellclick={[Function]}
  />
  <div />
  <Overlay
    displayText="WIN"
  />
</div>
`;
